lib/Dialect/ESI/runtime/cpp/lib/Manifest.cpp

```diff
@@ -0,0 +1,225 @@
+//===- Manifest.cpp - Metadata on the accelerator -------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// DO NOT EDIT!
+// This file is distributed as part of an ESI package. The source for this file
+// should always be modified within CIRCT
+// (lib/dialect/ESI/runtime/cpp/lib/backends/Cosim.cpp).
+//
+//===----------------------------------------------------------------------===//
+
+#include "esi/Manifest.h"
+#include "esi/Accelerator.h"
+#include "esi/Design.h"
+#include "esi/StdServices.h"
+
+#include <nlohmann/json.hpp>
+
+using namespace esi;
+
+namespace esi {
+namespace internal {
+
+// This is a proxy class to the manifest JSON. It is used to avoid having to
+// include the JSON parser in the header. Forward references don't work since
+// nlohmann::json is a rather complex template.
+//
+// Plus, it allows us to hide some implementation functions from the header
+// file.
+class ManifestProxy {
+public:
+  ManifestProxy(std::string jsonManifest, Manifest &manifest);
+
+  auto at(std::string key) const { return manifestJson.at(key); }
+
+  // Get the module info (if any) for the module instance in 'json'.
+  std::optional<ModuleInfo> getModInfo(const nlohmann::json &json) const;
+
+  // Build the 'Instance' recursively for the instance in 'json'.
+  Instance getInstance(const nlohmann::json &json) const;
+
+  /// Build the set of child instances for the module instance in 'json'.
+  std::vector<std::unique_ptr<Instance>>
+  getChildInstances(Accelerator &acc, const nlohmann::json &json) const;
+
+  /// Build a dynamic API for the Accelerator connection 'acc' based on the
+  /// manifest stored herein.
+  std::unique_ptr<Design> buildDesign(Accelerator &acc) const;
+
+private:
+  // The parsed json.
+  nlohmann::json manifestJson;
+  // Cache the module info for each symbol.
+  std::map<std::string, ModuleInfo> symbolInfoCache;
+};
+} // namespace internal
+} // namespace esi
+
+//===----------------------------------------------------------------------===//
+// Simple JSON -> object parsers.
+//===----------------------------------------------------------------------===//
+
+static AppID parseID(const nlohmann::json &json) {
+  std::optional<uint32_t> idx;
+  if (json.contains("index"))
+    idx = json.at("index").get<uint32_t>();
+  return AppID{json.at("name").get<std::string>(), idx};
+}
+
+static ModuleInfo parseModuleInfo(const nlohmann::json &mod) {
+  auto getAny = [](const nlohmann::json &json) -> std::any {
```

Comment:
one could blame this on `nlohmann::json` for naming a class in lowercase, but i don't like that the variable `json` alias with that. Suggesting rename to e.g. `value`.